qemu/tests/qemu_guest_agent.py

```diff
@@ -1918,8 +1946,129 @@ def gagent_check_fsfreeze(self, test, params, env):
+         one invalid mountpoint.
+        9) Check the result.
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environmen.
+        """
+        session = self._get_session(params, self.vm)
+        self._open_session_list.append(session)
+        image_size_stg0 = params["image_size_stg0"]
 
-        self._action_after_fsthaw(write_cmd, write_cmd_timeout)
+        error_context.context("Format the new data disk and mount it.",
+                              logging.info)
+        if params.get("os_type") == "linux":
+            disk_data = list(utils_disk.get_linux_disks(session).keys())
+            mnt_point_data = utils_disk.configure_empty_disk(
+                session, disk_data[0], image_size_stg0, "linux",
+                labeltype="msdos")[0]
+        else:
+            disk_index = utils_misc.wait_for(
+                lambda: utils_disk.get_windows_disks_index(session,
+                                                           image_size_stg0),
+                120)
+            if disk_index:
+                logging.info("Clear readonly for disk and online it in"
+                             " windows guest.")
+                if not utils_disk.update_windows_disk_attributes(session,
+                                                                 disk_index):
+                    test.error("Failed to update windows disk attributes.")
+                mnt_point_data = utils_disk.configure_empty_disk(
+                    session, disk_index[0], image_size_stg0, "windows",
+                    labeltype="msdos")[0]
+            else:
+                test.error("Didn't find any disk_index except system disk.")
+
+        if params.get("os_type") == "windows":
+            mount_points = ["C:\\", "%s:\\" % mnt_point_data]
+        else:
+            mount_points = ["/", mnt_point_data]
+
+        error_context.context("Freeze fs without specified mountpoints.",
+                              logging.info)
+        self._fsfreeze(fsfreeze_list=True, check_mountpoints=mount_points)
+        error_context.context("Freeze fs with two mount point.",
+                              logging.info)
+        self._fsfreeze(fsfreeze_list=True, mountpoints=mount_points)
+        error_context.context("Freeze fs with every mount point.",
+                              logging.info)
+        for mpoint in mount_points:
+            mpoint = ["%s" % mpoint]
+            self._fsfreeze(fsfreeze_list=True, mountpoints=mpoint)
+
+        error_context.context("Freeze fs with one valid mountpoint and"
+                              " one invalid mountpoint.", logging.info)
+        if params.get("os_type") == "linux":
+            mount_points_n = ["/", "/invalid"]
+            check_mp = ["/"]
+            self._fsfreeze(fsfreeze_list=True, mountpoints=mount_points_n,
+                           check_mountpoints=check_mp)
+        else:
```

Comment:
Would you help explain why windows can not directly share `_fsfreeze()` operations?

Reply:
For windows guest,need to catch the specified error info, if take it into _fsfreeze function,the code will more complex.

Reply:
@xiagao Okay, get it.